### PR TITLE
[Merged by Bors] - feat: port Data.Multiset.Fintype

### PR DESCRIPTION
**Pull request by int-y1**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by int-y1** [comment]
TODO: Fix the types for `Multiset.map_univ_coe` `Multiset.map_univ` `Multiset.prod_eq_prod_coe`

**Comment by Komyyy** [comment]
I've posted [the topic](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Data.2EMultiset.2EFintype.204!.231828) on Zulip.

File: Mathlib/Data/Multiset/Fintype.lean [hunk 1]
```diff
@@ -236,55 +232,56 @@ theorem Multiset.image_toEnumFinset_fst (m : Multiset α) :
...
 @[to_additive]
-theorem Multiset.prod_eq_prod_coe [CommMonoid α] (m : Multiset α) : m.Prod = ∏ x : m, x := by
+theorem Multiset.prod_eq_prod_coe [CommMonoid α] (m : Multiset α) : m.prod = ∏ x : m, (x : α) := by
   congr
-  simp
+  -- Porting note: `simp` fails with "maximum recursion depth has been reached"
+  erw [map_univ_coe]
 #align multiset.prod_eq_prod_coe Multiset.prod_eq_prod_coe
-#align multiset.sum_eq_sum_coe Multiset.sum_eq_sum_coe
+#noalign multiset.sum_eq_sum_coe
```
Comment (jcommelin) [inline Mathlib/Data/Multiset/Fintype.lean:269]: Why are you `noalign`ing these? they should be aligned to the additive version, as mathport suggests.

File: Mathlib/Data/Multiset/Fintype.lean [hunk 2]
```diff
@@ -236,55 +232,56 @@ theorem Multiset.image_toEnumFinset_fst (m : Multiset α) :
...
 
 @[to_additive]
 theorem Multiset.prod_eq_prod_toEnumFinset [CommMonoid α] (m : Multiset α) :
-    m.Prod = ∏ x in m.toEnumFinset, x.1 := by
+    m.prod = ∏ x in m.toEnumFinset, x.1 := by
   congr
   simp
 #align multiset.prod_eq_prod_to_enum_finset Multiset.prod_eq_prod_toEnumFinset
-#align multiset.sum_eq_sum_to_enum_finset Multiset.sum_eq_sum_toEnumFinset
+#noalign multiset.sum_eq_sum_to_enum_finset
```
Comment (jcommelin) [inline Mathlib/Data/Multiset/Fintype.lean:277]: Idem

File: Mathlib/Data/Multiset/Fintype.lean [hunk 3]
```diff
@@ -236,55 +232,56 @@ theorem Multiset.image_toEnumFinset_fst (m : Multiset α) :
...
-    simp
-  · simp
+  rw [Fintype.prod_equiv m.coeEquiv (fun x ↦ f x x.2) fun x ↦ f x.1.1 x.1.2]
+  · rw [← m.toEnumFinset.prod_coe_sort fun x ↦ f x.1 x.2]
+  · intro x
+    rfl
 #align multiset.prod_to_enum_finset Multiset.prod_toEnumFinset
-#align multiset.sum_to_enum_finset Multiset.sum_to_enum_finset
-
+#noalign multiset.sum_to_enum_finset
```
Comment (jcommelin) [inline Mathlib/Data/Multiset/Fintype.lean:287]: Idem

**Review by jcommelin — CHANGES_REQUESTED** [review]

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**